### PR TITLE
Implemented model_evaluation

### PR DESCRIPTION
**Pull request by Ahmad-M-Al-Khateeb**
I have implemented the model evaluation function.
To re-use some of the code that was originally in the trainer class, I have refactored it a bit.

Now most of the `compute_rewards` function is handled by a util function, as I needed that part for evaluation.

and I have transferred `prepare_first_attempt_input` and `prepare_second_attempt_input`   methods to the policy model class. Again because I would rather not re-write them for the evaluation.

Please let me know if there are any issues

### CODE REVIEW

**Comment by Ahmad-M-Al-Khateeb** [comment]
I will work on saving/loading models for now and once that's done I will actually test evaluation_model on a base model

**Review by AnkitXP — CHANGES_REQUESTED** [review]
Please enable Trainer rewards gradient and move to the same device as the model.

`rewards_tensor = torch.tensor(rewards, dtype=torch.float32, requires_grad=True).unsqueeze(1)        
        return rewards_tensor.to(self.policy_model.device)`

**Review by AnkitXP — APPROVED** [review]
